File: leetcode/python/LEET297.py
from typing import *
import collections

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        q = collections.deque([root])        
        res = ["#"]
        while q:
            node = q.popleft()
            if node:
                q.append(node.left)
                q.append(node.right)

                res.append(str(node.val))
            else:
                res.append("#")
        
        return ' '.join(res)
                
    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        if data == "# #":
            return None

        nodes = data.split()
        root = TreeNode(int(nodes[1]))
        q = collections.deque([root])
        index = 2

        while q:
            node = q.popleft()
            if nodes[index] is not "#":
                node.left = TreeNode(int(nodes[index]))
                q.append(node.left)
            index += 1

            if nodes[index] is not "#":
                node.right = TreeNode(int(nodes[index]))
                q.append(node.right)
            index += 1
        return root

File: leetcode/java/LEET297.java
/**
 * LEET297
 */
import java.util.*;
public class LEET297 {
    public static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
    public static class Codec {

        public String serialize(TreeNode root) {
            StringBuilder sb = new StringBuilder("# ");
            Queue<TreeNode> q = new LinkedList<>();
            q.add(root);

            while(!q.isEmpty()){
                TreeNode node = q.poll();

                if(node != null){
                    q.add(node.left);
                    q.add(node.right);

                    sb.append(node.val);
                } else {
                    sb.append("#");
                }
                sb.append(" ");
            }

            return sb.toString().trim();
        }
    
        // Decodes your encoded data to tree.
        public TreeNode deserialize(String data) {
            if(data.equals("# #")){
                return null;
            }
            Queue<TreeNode> q = new LinkedList<>();
            String[] nodes = data.split(" ");
            TreeNode root = new TreeNode(Integer.parseInt(nodes[1]));
            int index = 2;
            q.add(root);

            while(!q.isEmpty()){
                TreeNode node = q.poll();
                if(!nodes[index].equals("#")){
                    node.left = new TreeNode(Integer.parseInt(nodes[index]));
                    q.add(node.left);
                }
                index++;

                if(!nodes[index].equals("#")){
                    node.right = new TreeNode(Integer.parseInt(nodes[index]));
                    q.add(node.right);
                }
                index++;
            }
            return root;
        }     // Encodes a tree to a single string.
   
    }
    
    
}